### PR TITLE
better error for missing pubkey + nix shell + circle

### PR DESCRIPTION
**Pull request by thedavidmeister**
i was finding the "bad type" error very cryptic when debugging downstream for the pubkey

i also setup nix and circle

### CODE REVIEW

**Comment by neonphog** [comment]
@thedavidmeister 

> i also setup nix

Why? This repo has nothing to do with the HDK - hApp devs will not need access to modify this code - even if they are more core oriented and want to help, I vote we just keep this as a simple canonical rust repo. There's no reason to take on all the nix overhead.

**Comment by neonphog** [comment]
@thedavidmeister 

> and circle

Why? This is a simple rust repo, it takes very little time to run on CI, there's no reason to pay for CI. I actually would like to switch away from travis and just use github actions, it will be even more integrated / easy to work with.

**Comment by neonphog** [comment]
@thedavidmeister 

As for integrating into Holonix - we should just commit the Cargo.lock in the binary crate, and use carnix from the other side - just like any other rust dependency we might want to bring into Holonix.

**Review by neonphog — CHANGES_REQUESTED** [review]
I like the error handling change, If you extract just that I'd be happy to merge.

**Comment by thedavidmeister** [comment]
> @thedavidmeister
> 
> > i also setup nix
> 
> Why? This repo has nothing to do with the HDK - hApp devs will not need access to modify this code - even if they are more core oriented and want to help, I vote we just keep this as a simple canonical rust repo. There's no reason to take on all the nix overhead.

@neonphog HDK and nix are orthogonal

nix standardises:

- dependencies
- release scripts
- development environment variables
- CI setup and deployment
- staff training, devices and ability to manage code

across all organisation repositories

i don't know what 'all the nix overhead' is so i can't directly respond to that

**Comment by thedavidmeister** [comment]
> @thedavidmeister
> 
> As for integrating into Holonix - we should just commit the Cargo.lock in the binary crate, and use carnix from the other side - just like any other rust dependency we might want to bring into Holonix.

@neonphog yeah that makes sense, i'll merge this in - https://github.com/holochain/lair/pull/15 - and based on our ops experience to date we should be smoke testing builds here (e.g. based on carnix) before we put up release candidates to be consumed downstream, otherwise we run into the issue where it becomes the problem of the person trying to integrate code (e.g. often myself) that is broken (e.g. windows doesn't compile because we haven't been testing it to date) and then they have to fix it and integrate it, which causes all kinds of unnecessary delays and stress

**Comment by thedavidmeister** [comment]
> @thedavidmeister
> 
> > and circle
> 
> Why? This is a simple rust repo, it takes very little time to run on CI, there's no reason to pay for CI. I actually would like to switch away from travis and just use github actions, it will be even more integrated / easy to work with.

@neonphog the pricing and tiers for circle and github actions seem to be pretty similar

i'm happy to try out github actions though, i've been experimenting with them here

**Comment by thedavidmeister** [comment]
@neonphog try now

File: .github/workflows/ops.test.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+   - uses: actions-rs/cargo@v1
+     with:
+       command: build
+       args: --release
+
+   # run all the tests
+   - uses: actions-rs/cargo@v1
+     with:
+       command: test
+       args: --release
```
Comment (neonphog) [inline .github/workflows/ops.test.yml:50]: Tests should not be run in release mode - when there are errors, especially flaky errors, we want to see the line numbers.

```suggestion
       args:
```

File: .github/workflows/bin_tests.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+   - uses: actions-rs/cargo@v1
+     with:
+       command: install
+       args: -f --path crates/lair_keystore
+
+   # run the binary tests
+   - uses: actions-rs/cargo@v1
+     with:
+       command: test
+       args: --release --manifest-path crates/lair_keystore_client/Cargo.toml --features=bin-tests
```
Comment (neonphog) [inline .github/workflows/bin_tests.yml:50]: ditto tests should not be run in release mode

```suggestion
       args: --manifest-path crates/lair_keystore_client/Cargo.toml --features=bin-tests
```

**Review by neonphog — CHANGES_REQUESTED** [review]
I'd still love it if these commands were executed through whatever standard commands we want to have in our repo -- i.e. `make clippy` or whatever, just so we only have to update things in one place - but this is cool until we sort out what we're doing with that.

Will approve with the `--release` removed from testing.

**Review by neonphog — APPROVED** [review]

**Comment by thedavidmeister** [comment]
> I'd still love it if these commands were executed through whatever standard commands we want to have in our repo -- i.e. `make clippy` or whatever, just so we only have to update things in one place - but this is cool until we sort out what we're doing with that.
> 
> Will approve with the `--release` removed from testing.

@neonphog yeah that's basically the plan, probably for a future PR, i made the yaml files composable so we could even have CI config standardised like that

**Pull request merged**